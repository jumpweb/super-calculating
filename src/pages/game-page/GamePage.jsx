import {useState} from "react";
import IntervalButton from './IntervalButton'
import GameBoard from "./game-board/GameBoard";
import GamePresenter from "./GamePresenter";
import ResultsBoard from "./ResultsBoard";
import InGameStatistics from "./game-staticstics/InGameStatistics";
import GameController from './game-controller/GameController';
import QuestionCountSelector from "./question-count-selector/QuestionCountSelector";

function GamePage({ defaultNumberOfQuestions, startupCountDown }) {
    const [numberOfQuestions, setNumberOfQuestions] = useState(defaultNumberOfQuestions)
    const [gameStarting, setGameStarting] = useState(false)

    return (
        <GameController numberOfQuestions={numberOfQuestions}>
            {props => (
                <GamePresenter
                    questionSelector={
                        <QuestionCountSelector
                            hide={gameStarting}
                            defaultNumberOfQuestions={defaultNumberOfQuestions}
                            onChange={setNumberOfQuestions}
                        />
                    }
                    gameState={props.gameState}
                    inGameStats={
                        <InGameStatistics
                            elapsedSeconds={props.elapsedSeconds}
                            questionNumber={props.numberOfQuestionsAsked}
                            totalNumberOfQuestions={numberOfQuestions}
                        />
                    }
                    startButton={
                        <IntervalButton
                            text={startupCountDown}
                            onCountDownStarted={() => setGameStarting(true)}
                            onCountDownCompleted={props.onStartGame}
                        />
                    }
                    gameBoard={
                        <GameBoard
                            hasMoreQuestions={props.hasMoreQuestions}
                            onAskNextQuestion={props.onAskNextQuestion}
                            onQuestionAnswered={props.onQuestionAnswered}
                            onEndGame={props.onEndGame}
                        />
                    }
                    scoreBoard={
                        <ResultsBoard
                            elapsedTime={props.elapsedSeconds}
                            totalNumberOfQuestions={numberOfQuestions}
                            numberOfQuestionsAnsweredCorrectly={props.numberOfQuestionsAnsweredCorrectly}
                            numberOfQuestionsAnsweredIncorrectly={props.numberOfQuestionsAnsweredIncorrectly}
                            answerHistory={props.answerHistory}
                        />
                    }
                />
            )}
        </GameController>
    )
}

export default GamePage
