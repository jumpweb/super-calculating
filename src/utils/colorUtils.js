function random(number){
    return Math.floor(Math.random()*number);;
}

export function getRandomColor() {
    return 'rgb('+random(255)+','+random(255)+','+random(255)+')';
}
