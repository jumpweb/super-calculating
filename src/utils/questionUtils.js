export const Question = {
    set: ({ value, answer }) => {
        return {
            value,
            answer,
        }
    },
    empty: () => {
        return {
            value: '',
            answer: 0
        }
    }
}

const TwoDigitAdditionQuestion = {
    generate: () => {
        const firstNumber = Math.floor(Math.random() * 10) + 1;
        const secondNumber = Math.floor(Math.random() * 10) + 1;

        return Question.set({
            value: `${firstNumber} + ${secondNumber}`,
            answer: firstNumber + secondNumber
        })
    }
}

export function generateQuestion() {
    return TwoDigitAdditionQuestion.generate()
}
