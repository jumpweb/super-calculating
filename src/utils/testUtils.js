import {fireEvent, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

export function answerTheQuestionWith(answerValue) {
    const answerInputField = screen.getByLabelText('Enter your answer')
    const checkAnswerButton = screen.getByText('Check Answer')

    fireEvent.change(answerInputField, {target: {value: answerValue}})
    userEvent.click(checkAnswerButton)
}
