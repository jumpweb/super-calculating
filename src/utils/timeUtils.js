export const formatElapsedSeconds = (elapsedSeconds) => {
    return new Date(elapsedSeconds * 1000)
        .toISOString()
        .substr(14, 5)
}
