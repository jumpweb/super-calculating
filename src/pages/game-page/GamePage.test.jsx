import {render, screen, act, waitFor, within, fireEvent} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as questionUtils from '../../utils/questionUtils'
import {TwoDigitQuestion} from "../../utils/questionUtils";
import {answerTheQuestionWith, clickTheEndGameButton, clickTheNextButton} from "../../utils/testUtils";
import GamePage from './GamePage'

jest.useFakeTimers()

jest.mock('@mui/material/Slider', () => (props) => {
    const { id, name, min, max, defaultValue, onChange } = props;
    return (
        <input
            aria-labelledby={props['aria-labelledby']}
            type="range"
            id={id}
            name={name}
            min={min}
            max={max}
            value={defaultValue}
            onChange={onChange}
        />
    );
});

describe('Game page', () => {
    describe('before starting the game', () => {
        it('Sets a default number of questions', async () => {
            renderGamePage()

            expect(screen.getByLabelText(/Number of questions:/).value).toEqual('3')
        })

        it('allows the player to select the number of questions from a predefined list', async () => {
            renderGamePage()

            fireEvent.change( screen.getByLabelText(/Number of questions:/), { target: { value: 2 } })

            await pressTheStartButton()

            await waitForQuestion()

            expect(screen.getByText('Question 1 of 2')).toBeInTheDocument()
        })

        it('asks only the selected number of questions before ending the game', async () => {
            renderGamePage()

            fireEvent.change( screen.getByLabelText(/Number of questions:/), { target: { value: 2 } })

            await pressTheStartButton()

            await waitForQuestion()

            await answerTheQuestionWith('13')
            await clickTheNextButton()

            await answerTheQuestionWith('9')

            expect(screen.getByText('End game')).toBeInTheDocument()
        })
    })

    describe('Playing the game', () => {
        beforeEach(startTheGame)

        describe('when the start button is pressed', () => {
            it('hides the question selector', () => {
                expect(screen.queryByLabelText(/Number of questions:/)).not.toBeInTheDocument()
            })

            it('Counts down from Start to 1', () => {
                expect(screen.getByText('1')).toBeInTheDocument()
            })

            it('Counts down from 1 to GO!', async () => {
                await waitForGo()

                expect(screen.getByText('GO!')).toBeInTheDocument()
            })

            it('displays a mathematics question when the count down reaches zero', async () => {
                await waitForQuestion()

                expect(screen.getByText('5 + 8 =')).toBeInTheDocument()
            })
        })

        describe('while playing the game', () => {
            it('displays the initial question number out of the total number of questions', async () => {
                await waitForQuestion()

                expect(screen.getByText('Question 1 of 3')).toBeInTheDocument()
            })

            it('displays the initial elapsed time as zero', async () => {
                await waitForQuestion()

                expect(screen.getByText('Elapsed time: 00:00')).toBeInTheDocument()
            })

            it('displays the elapsed number of seconds since the game started', async () => {
                await waitForQuestion()

                act(() => {
                    jest.advanceTimersByTime(2000)
                })

                expect(screen.getByText('Elapsed time: 00:02')).toBeInTheDocument()
            })

            it('pauses the timer while displaying the result of the answer', async () => {
                await waitForQuestion()

                act(() => {
                    jest.advanceTimersByTime(2000)
                })

                await answerTheQuestionWith('10')

                act(() => {
                    jest.advanceTimersByTime(3000)
                })

                await clickTheNextButton()

                act(() => {
                    jest.advanceTimersByTime(1000)
                })

                expect(screen.getByText('Elapsed time: 00:03')).toBeInTheDocument()
            })

            it('displays the updated question number out of the total number of questions', async () => {
                await waitForQuestion()

                await answerTheQuestionWith('10')
                await clickTheNextButton()

                expect(screen.getByText('Question 2 of 3')).toBeInTheDocument()
            })
        })
    })

    describe('after the game has ended', () => {
        it('displays the score board when all the questions have been answered', async () => {
            const { container } = await startTheGame()

            await waitForQuestion()

            act(() => {
                jest.advanceTimersByTime(10000)
            })

            await answerTheQuestionWith('13')
            await clickTheNextButton()

            await answerTheQuestionWith('9')
            await clickTheNextButton()

            await answerTheQuestionWith('10')
            await clickTheEndGameButton()

            expect(container.firstChild).toMatchSnapshot()
        })
    })
})

async function startTheGame() {
    const renderResult = await renderGamePage()

    pressTheStartButton()

    return renderResult
}

function renderGamePage() {
    jest.spyOn(questionUtils, 'generateQuestion')
        .mockReturnValue(TwoDigitQuestion.set(5, 8))

    return render(<GamePage defaultNumberOfQuestions={3} startupCountDown={['Start', '1', 'GO!']} />)
}

function pressTheStartButton() {
    userEvent.click(screen.getByText('Start'))
}

const waitForGo = waitForCountdown(1000)
const waitForQuestion = waitForCountdown(2000)

function waitForCountdown(period) {
    return async function() {
        await waitFor(() => {
            jest.advanceTimersByTime(period);
        })
    }
}
