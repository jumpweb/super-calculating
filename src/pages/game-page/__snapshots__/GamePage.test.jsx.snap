// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game page after the game has ended displays the score board when all the questions have been answered 1`] = `
<div
  class="css-p192ua"
>
  <div
    class="css-1g4yje1"
  >
    <h3
      class="css-i51og3"
    >
      It's the end of the game, here are your results.
    </h3>
    <table
      class="MuiTable-root css-sumxgb-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1y5vm7s-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1howxi1-MuiTableCell-root"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-12q85cb-MuiTableCell-root"
            scope="col"
          >
            Current game
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall css-12q85cb-MuiTableCell-root"
            scope="col"
          >
            Previous game
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1y5vm7s-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Time taken
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            00:10
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            02:00
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1y5vm7s-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Questions asked
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            10
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1y5vm7s-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Questions answered correctly
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            2
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1y5vm7s-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Questions answered incorrectly
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            8
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1y5vm7s-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
          >
            Percentage answered correctly
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            33.33333333333333%
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall css-dv0mq-MuiTableCell-root"
          >
            20%
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="MuiTableContainer-root css-6gfez-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeSmall css-nc6t7a-MuiTableCell-root"
              scope="col"
            >
              Question
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeSmall css-nc6t7a-MuiTableCell-root"
              scope="col"
            >
              You answered
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeSmall css-nc6t7a-MuiTableCell-root"
              scope="col"
            >
              Correct answer
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              5 + 8
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              13
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              13
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              <p>
                ✅
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              5 + 8
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              9
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              13
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              <p>
                ❌
              </p>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              5 + 8
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              10
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              13
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall css-vtdehq-MuiTableCell-root"
            >
              <p>
                ❌
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
